### PR TITLE
feat: NF41 Speedrun (Arduino)

### PR DESCRIPTION
**Pull request by zfields**

### CODE REVIEW

File: 41-vaccine-temp-monitor/README.md [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+Notehub Setup
+-------------
+
+Sign up for a free account on [notehub.io](https://notehub.io) and
+[create a new project](https://dev.blues.io/quickstart/notecard-quickstart/notecard-and-notecarrier-pi/#set-up-notehub).
+
+### Twilio Route
+
+When an alert condition is detected a Note is immediately sent to Notehub. In
+order for you to be informed this event occurred, a Twilio route has been
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/README.md:67]: I would rephrase this. Just reads weird with passive voice. Suggestion:

In order for you to be informed this event occurred, you must configure a Twilio route.

File: 41-vaccine-temp-monitor/README.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+enabled.
+
+Follow the
+[Twilio SMS Guide on https://blues.dev](https://dev.blues.io/guides-and-tutorials/twilio-sms-guide/)
+to learn the steps required to link your Twilio account to Notehub.
+
+Once you are familiarized with the Twilio route, then create a new route and set
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/README.md:74]: familiarized -> familiar
then create -> create

File: 41-vaccine-temp-monitor/README.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+Environment Variables
+---------------------
+
+- `alarm_temperature` - The temperature at which the alarm should trigger.
+- `alarm_hysteresis` - The temperature at which the alarm should reset.
+- `sample_interval_s` - The temperature and battery sampling interval in seconds.
+
+_**NOTE:** If you modify any environment variables, you must wait until the next
+synchronization event (max wait: inbound frequency) for the changes to propagate
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/README.md:214]: I don't think newbs are going to understand what you mean by

> max wait: inbound frequency

Worth explaining.

EDIT: I see you go into this a bit more below, so you could also just mention here that you'll elaborate later in the document.

File: 41-vaccine-temp-monitor/README.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+Normal data:
+
+```json
+{
+    "temp":18.5,
+    "voltage":3.5898438
+}
+```
+
+#### Location sampling
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/README.md:256]: sampling -> Sampling

File: 41-vaccine-temp-monitor/README.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+
+The device is configured to use motion based GPS tracking. When motion is
+detected, the device will report its location periodically; every 15 minutes.
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/README.md:259]: I find this sentence hard to understand. Does it report when motion is detected, or every 15 minutes? How can it be both? Maybe something like this would be better:

As long as motion has been detected, the device will report its location once every 15 minutes.

File: 41-vaccine-temp-monitor/README.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+##### Thermal Event Operation
+
+During a thermal event, time is of the essense and the mode of operation changes
+accordingly. Instead of sampling every 30 minutes, the temperature and location
+will be sampled every every 3 minutes. While in this mode, the device will force
+a GPS fix or timeout after 95 seconds.
+
+#### Low Battery
+
+When `card.voltage` returns `"mode":"low"`, then the `low_batt` key is added to
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/README.md:298]: What is card.voltage? I think we should assume the reader doesn't know anything about Notecard requests.

File: 41-vaccine-temp-monitor/README.md [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+    "voltage":3.5898438
+}
+```
+
+When the `low_batt` key is received by Notehub, then an SMS message containing
+the voltage is forwared to Twilio. Again, a text message is generated and sent
+to the phone number specified in the `alarm_sms_number` environment variable.
+
+_**NOTE:** The low battery alert IS NOT considered to be a high-priority alert
+and will be sent at the regular sync'ing interval._
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/README.md:316]: sync'ing -> syncing

**Review by haydenroche5 — CHANGES_REQUESTED** [review]
Review comments for README.md. Will do firmware next.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+// #define PRODUCT_UID "com.your-company:your-product-name"
+
+#ifndef PRODUCT_UID
+#define PRODUCT_UID "" // "com.my-company.my-name:my-project"
+#pragma message "PRODUCT_UID is not defined in this example. Please ensure your Notecard has a product identifier set before running this example or define it in code here. More details at https://bit.ly/product-uid"
+#endif
+#define usbSerial Serial
+#define txRxPinsSerial Serial1
+
+#define CRT_MAX_TEMP_C 30
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:45]: What's CRT?

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 9]
```diff
@@ -0,0 +1,528 @@
...
+    va_list args;
+    va_start(args, format_);
+    vsnprintf(log, sizeof(log), format_, args);
+
+    if (J *cmd = notecard.newCommand("hub.log"))
+    {
+        JAddStringToObject(cmd, "text", log);
+        if (!notecard.sendRequest(cmd))
+        {
+            JDelete(cmd);
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:74]: Doesn't sendRequest always delete the passed in request, even if it fails?

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 10]
```diff
@@ -0,0 +1,528 @@
...
+{
+    int result;
+    const char *gps_mode;
+    size_t gps_time_s;
+
+    // Get Current Configuration
+    if (J *rsp = notecard.requestAndResponse(notecard.newRequest("card.location.mode")))
+    {
+        gps_mode = JGetString(rsp, "mode");
+        NoteDeleteResponse(rsp);
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:91]: Doesn't deleting the response invalidate the gps_mode pointer? I think you would need to copy the string out before deleting.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 11]
```diff
@@ -0,0 +1,528 @@
...
+    {
+        result = -1; // Unable to communicate with Notecard
+        return result;
+    }
+
+    // Halt periodic tracking, before active sync is performed
+    if (J *cmd = notecard.newCommand("card.location.mode"))
+    {
+        JAddStringToObject(cmd, "mode", "off");
+        notecard.sendRequest(cmd);
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:103]: Should we check the return here?

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 12]
```diff
@@ -0,0 +1,528 @@
...
+    {
+        result = -1; // Unable to communicate with Notecard
+        return result;
+    }
+
+    // Activate GPS
+    if (J *cmd = notecard.newCommand("card.location.mode"))
+    {
+        JAddStringToObject(cmd, "mode", "continuous");
+        notecard.sendRequest(cmd);
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:123]: Same question about checking return.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 13]
```diff
@@ -0,0 +1,528 @@
...
+    }
+
+    // Block while resolving GPS location
+    for (const size_t start_ms = ::millis();;)
+    {
+        if (timeout_s_)
+        {
+            if (::millis() >= (start_ms + (timeout_s_ * 1000)))
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:131]: I think this has the millis rollover problem. https://arduino.stackexchange.com/questions/12587/how-can-i-handle-the-millis-rollover

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 14]
```diff
@@ -0,0 +1,528 @@
...
+
+        // Check if GPS has acquired location information
+        if (J *rsp = notecard.requestAndResponse(notecard.newRequest("card.location")))
+        {
+            const size_t current_gps_time_s = JGetInt(rsp, "time");
+            if (current_gps_time_s != gps_time_s)
+            {
+                logNoteF("INFO: Current GPS time (seconds): %u", current_gps_time_s);
+                result = 0; // GPS has fixed on position
+                break;
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:147]: This leaks rsp, I think. Same below with the other break.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 15]
```diff
@@ -0,0 +1,528 @@
...
+            break;
+        }
+        ::delay(2500);
+    }
+
+    // Restore Previous Configuration
+    if (J *cmd = notecard.newCommand("card.location.mode"))
+    {
+        JAddStringToObject(cmd, "mode", gps_mode);
+        notecard.sendRequest(cmd);
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:170]: Check return.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 16]
```diff
@@ -0,0 +1,528 @@
...
+    return result;
+}
+
+void configureNotecard(void)
+{
+    // Configure Notecard to synchronize with Notehub periodically, as well as
+    // adjust the frequency based on the battery level
+    if (J *req = notecard.newCommand("hub.set"))
+    {
+        JAddStringToObject(req, "sn", "EpiCAT");
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:183]: We should add the app name to the user agent, too. See https://github.com/blues/app-accelerators/pull/113.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 17]
```diff
@@ -0,0 +1,528 @@
...
+        if (PRODUCT_UID[0])
+        {
+            JAddStringToObject(req, "product", PRODUCT_UID);
+        }
+        JAddStringToObject(req, "mode", "periodic");
+        JAddStringToObject(req, "vinbound", "usb:5;high:120;normal:240;low:480;dead:0");
+        JAddStringToObject(req, "voutbound", "usb:5;high:60;normal:120;low:240;dead:0");
+        notecard.sendRequestWithRetry(req, RETRY_S);
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:191]: Check return. I'll stop saying it here, but there are more below. :)

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 18]
```diff
@@ -0,0 +1,528 @@
...
+
+    // Restore motion sensitivity to preserve battery life
+    if (J *cmd = notecard.newCommand("card.motion.mode"))
+    {
+        JAddBoolToObject(cmd, "start", true);
+        JAddNumberToObject(cmd, "sensitivity", 1);
+        notecard.sendRequest(cmd);
+    }
+
+    return valid;
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:272]: `valid` not declared. This function seems incomplete.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 19]
```diff
@@ -0,0 +1,528 @@
...
+}
+
+void envVarManagerCb(const char *var, const char *val, void *userCtx)
+{
+    // Cache the values for each variable.
+    if (strcmp(var, "alarm_temperature") == 0)
+    {
+        alarm_temperature = atoi(val);
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:280]: You may want to add some sanity checks on these values. e.g. make sure it's actually a number. Otherwise, atoi is going to return 0 on error, and you'll start to get some wacky alerts.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 20]
```diff
@@ -0,0 +1,528 @@
...
+#ifdef ARDUINO_ARCH_ESP32
+    // Initialize Auxiliary Wi-Fi
+    aux_wifi.begin();
+#endif
+
+    // Configure Notecard
+    configureNotecard();
+
+    // Fetch Environment Variables
+    fetchEnvironmentVariables();
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:394]: So we only fetch environment variables at startup? Just checking that's the intent. No way to update the values without rebooting.

EDIT: Or does setup get run again every time we wake up? That may be worth clarifying with a comment, if so.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 21]
```diff
@@ -0,0 +1,528 @@
...
+
+    // Sample battery voltage
+    float voltage = 0.0f;
+    bool low_battery = false;
+    if (J *rsp = notecard.requestAndResponse(notecard.newRequest("card.voltage")))
+    {
+        voltage = JGetNumber(rsp, "value");
+        const char *power_mode = JGetString(rsp, "mode");
+        low_battery = (power_mode && !strcmp(power_mode, "low"));
+        NoteDeleteResponse(rsp);
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:424]: We're using both NoteDeleteResponse and notecard.deleteResponse in this code. Probably best to settle on one or the other.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 22]
```diff
@@ -0,0 +1,528 @@
...
+    // Configure Notecard
+    configureNotecard();
+
+    // Fetch Environment Variables
+    fetchEnvironmentVariables();
+
+    // Configure temperature threshold for Controlled Room Temperature (CRT)
+    // U.S. Pharmacopeia (USP) USP <659> "Packaging and Storage Requirements"
+    //
+    // Controlled room temperature: The temperature maintained thermostatically
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:399]: This may be better to have toward the top where you first use the term CRT.

**Review by haydenroche5 — CHANGES_REQUESTED** [review]

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp [hunk 23]
```diff
@@ -0,0 +1,528 @@
...
+void fetchEnvironmentVariables(void)
+{
+    static const char *envVars[] = {
+        "alarm_temperature",
+        "alarm_hysteresis",
+        "sample_interval_s"};
+    static const size_t numEnvVars = sizeof(envVars) / sizeof(envVars[0]);
+
+    // Allocate the environment variable manager.
+    if (!(envVarManager = NotecardEnvVarManager_alloc()))
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:301]: Does all the memory get wiped out on sleep? Just trying to figure out why we can continually alloc the manager and never free it.

Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:301]: Yes.

BUT... I would agree this is not a very tidy thing to do.

I'll add `NotecardEnvVarManager_free()` to clean up after myself.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 21 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:424]: `NoteDeleteResponse()` snuck in there. It's dead now :skull:

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 20 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:394]: Yes, due the nature of the application, it's a once and done. I will add a comment to this affect.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 19 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:280]: good thinking!

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 20 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:394]: After having seen it in context, I think the comment belongs in `README.md`.

Otherwise, this "problem" would plague nearly every aspect of this type of application.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 19 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:280]: I appended `&& atoi(val)` to each `if` statement.

Tell me if that's too gross.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 8 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:45]: You don't know!? I heard they were teaching it in schools!

Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:45]: Code comment moved in accordance with your subsequent review comment.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 9 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:74]: No.

Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:74]: It didn't used to, but apparently it does now. :joy:

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 10 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:91]: Great catch! Absolutely right.

Not sure what I was thinking. If it was making a copy, then I would have been leaking. :shrug:

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 18 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:272]: Yeah, I don't know what I pushed up exactly, that became `moving` and subsequently made sense. :shrug:

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 12 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:123]: I'm of two minds on this, and I'll chat with you in person tomorrow.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 11 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:103]: answered below

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 19 above)
Comment (haydenroche5) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:320]: That's fine, but it's silent if the conversion fails. I think it'd be nice to get a debug log saying the conversion failed and that we're keeping the old value.

Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:320]: you would...

File: 41-vaccine-temp-monitor/README.md (hunk 5 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/README.md:259]: Tried to clean it up by shuffling words around. Let me know what you think.

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 19 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp]: I updated it to log the failure to parse to Notehub, so the log will appear next to the environment variables (where they can actually do something about it).

File: 41-vaccine-temp-monitor/README.md [hunk 24]
```diff
@@ -0,0 +1,350 @@
...
+Normal data:
+
+```json
+{
+    "temp":18.5,
+    "voltage":3.5898438
+}
+```
+
+_**NOTE:** Temperature is rendered in Celcius._
```
Comment (haydenroche5) [inline 41-vaccine-temp-monitor/README.md:275]: Celcius -> Celsius

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 15 above)
Comment (haydenroche5) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp]: We decided offline that checking a lot of these isn't super necessary. Any errors will come back from the Notecard's debug logging (after changing from commands to requests).

**Review by haydenroche5 — CHANGES_REQUESTED** [review]

File: 41-vaccine-temp-monitor/README.md (hunk 24 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/README.md:275]: You and your dang comment... :joy:

File: 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp (hunk 14 above)
Comment (zfields) [reply 41-vaccine-temp-monitor/firmware/arduino/src/main.cpp:163]: I was demonstrating how you can leak memory and it just automatically gets cleaned up when the Notecard controls the power supply of the host. :rofl:

**Review by haydenroche5 — APPROVED** [review]

**Comment by m-mcgowan** [comment]
[NF41 - Vaccine Temp Monitoring](https://trello.com/c/zpCkhYGK/209-nf41-vaccine-temp-monitoring)
